cljfmt/src/cljfmt/core.cljc

```diff
@@ -702,6 +709,24 @@
       (edit-all non-indenting-whitespace? unalign-from-space z/right*)
       z/root))
 
+(defn- blank-line-in-form? [zloc blank-line-forms context]
+  (and (z/linebreak? zloc)
+       (> (count-newlines zloc) 1)
+       (some (fn [zloc*]
+               (and (z/list? (z/up zloc*))
+                    (not (matching-form? zloc* blank-line-forms context))))
+             [zloc (z/up zloc)])))
```

Comment:
Could you explain why we need to check `(z/up zloc)` and `(z/up (z/up zloc))`?

Reply:
`(z/up zloc)`  ends up matching `:all` rules in cases like this 

```clj
(cond
  a
  b
  <this-node-right-here>
  c
  d)
```

while `(z/up (z/up zloc))`  ends up matching the set-of-indexes rules (e.g. `#{0}`) in cases like this

```clj
(let [a
      1
      <this-node-right-here>
      b
      2]
  ...)
```

In the first case the form that the rules match (`cond`) is the parent of the newline node while in the second case the form (`let`) is the grandparent of the newline node.

I can move the logic for checking at two different levels into `matching-form?` itself if you'd like. I was trying to make things work using your suggested changes exactly but I think this change would make things cleaner